### PR TITLE
Support asynchronous `IO.pipe` on Windows

### LINKED ISSUE #6957: Port concurrency features to Windows
This is a sub-task of porting the stdlib to Windows #5430

* [ ] Extract platform-specifics of `Fiber` (#6955)
* [ ] Refactor and port `Crystal::Event`, `Crystal::EventLoop`
* [ ] Port `Crystal::System::Fiber`
* [ ] [Port thread handling (`Thread`, `Thread::Mutex`, `GC`)]

I suppose we can delay porting threads by implementing a mock API for `Thread` and `Thread::Mutex` for win32 which essentially doesn't to anything. That should work perfectly fine for single threaded process.

On windows, we should use the win32 API directly instead of libevent (quoting @RX14):

> libevent uses select on windows, and select doesnt support pipes on windows, only sockets. Besides, it's an additional dependency, whereas using IOCP directly isn't that much harder and allows us to support sockets, pipes, files, etc.
> Windows's nonblocking IO model is entirely different to how libevent works. On windows you submit entire IO operations and wait for them to complete. With epoll you're waiting for "readable/writable status" of a FD then you can perform an IO.

Since the API models are quite different, this will likely require some refactoring of  `Crystal::Event` and `Crystal::EventLoop`.

### LINKED ISSUE #12718: Process.run hangs on windows in conditions
Hi there, 

I had posted a comment on the crystal-lang [forums](https://forum.crystal-lang.org/t/process-run-on-windows/5051) with this issue. It looks to be a separate issue from #12241. 
As far as I can tell, running a process with the exact same arguments twice in a conditional block will cause crystal to hang. 

Here is a reproduction that runs in linux, but hangs in windows. 

```
#>crystal -v
#Crystal 1.7.0-dev [1c33a0b] (2022-10-29)

#LLVM: 13.0.0
#Default target: x86_64-pc-windows-msvc

def run_proc(command, a)
  stdout = IO::Memory.new
  stderr = IO::Memory.new
  begin
    mini_proc = Process.run(command, args: a, output: stdout, error: stderr)
    if mini_proc.success?
      return stdout.to_s
    else
      return "error" #stderr.to_s
    end
  rescue ex
    return "error"
  end
end

# This part works fine on windows
puts "windows version"
puts run_proc("wmic", ["os", "get", "Caption", "/value"]).strip

# This part stalls after 'try wmic' and never prints the 'acting on ... ' messages
os_info = ""
puts "try uname"
if run_proc("uname", ["-a"]) == "error" # linux/mac/bsd fail
    puts "try wmic"
    if run_proc("wmic", ["os", "get", "Caption", "/value"]) == "error" # windows fail
        puts "acting on wmic fail"
        os_info = "Unknown"
    else # windows success
        puts "acting on wmic success"
        os_info = run_proc("wmic", ["os", "get", "Caption", "/value"]).strip
    end
else # linux/mac/bsd success
    os_info = run_proc("uname", ["-a"])
end

puts os_info

```

### PR DESCRIPTION
**Pull request by HertzDevil**
This PR makes `IO.pipe` return `IO::FileDescriptor`s whose `#read` and `#write` run asynchronously. Closes #6957. Fixes #12718. Fixes https://github.com/crystal-lang/crystal/pull/9893#discussion_r522502821.

`IO.pipe` is backed by Win32 named pipes with unique names. They differ from Unix pipes in at least two substantial ways:

* One end is the pipe server and the opposite end is the pipe client, even when both ends are only used in the same process. The pipe direction is independent of the server-client relationship; this PR makes the write end the server end.
* Closing the write end does not signal an EOF on the read end, but it does trigger `ERROR_BROKEN_PIPE`, so that error value is used for this purpose instead and not treated as a hard error.

Asynchronous I/O for regular files might be as simple as passing `LibC::FILE_FLAG_OVERLAPPED` to `LibC.CreateFileW`, which I haven't tried. (They will signal EOFs.) The console `STDIN` and `STDOUT` are also still synchronous and might require separate threads as described in https://github.com/crystal-lang/crystal/issues/6957#issuecomment-748569797.

Also it seems this is not enough to get the playground running on Windows.

### CODE REVIEW

File: spec/std/http/spec_helper.cr [hunk 1]
```diff
@@ -61,29 +61,26 @@ end
...
 
-      begin
-        wait_until_blocked f
+    begin
+      wait_until_blocked f
 
-        yield client
-      ensure
-        processor.close
-        server_io.close
```
Comment (HertzDevil) [inline spec/std/http/spec_helper.cr:83]: It seems one of the `IO`s was trying to flush to a closed stream if this line is present. Either way the block form of `IO::Stapled.pipe` automatically closes the two `IO`s so this line shouldn't be necessary here

File: spec/std/http/spec_helper.cr [hunk 2]
```diff
@@ -61,29 +61,29 @@ end
...
-        processor.close
+      yield client
+    ensure
+      processor.close
+      {% unless flag?(:win32) %}
         server_io.close
-        if exc = done.receive
-          raise exc
-        end
+      {% end %}
```
Comment (HertzDevil) [inline spec/std/http/spec_helper.cr:84]: It seems one of the `IO`s was trying to flush to a closed stream if this line is present. I have no idea this broke the OAuth2 specs

**Comment by straight-shoota** [comment]
My main worry is the behaviour of async (overlapped) streams being passed to a new process. As described in https://github.com/crystal-lang/crystal/issues/6957#issuecomment-748569797 point 3. this would be problematic.
This issues does not appear with most typical process IO. The pipes created by `Process#stdio_to_fd` are blocking on the end that goes to the new process.
But when you explicitly pass a file descriptor that's non-blocking (such as one created by `IO.pipe`), there may be problems.

The following program has lead to crashes with this patch:
```cr
reader, writer = IO.pipe
Process.run("cmd.exe", output: writer)
writer.close
puts reader.gets_to_end
```
It doesn't always crash. And I haven't been able to dig more into what's going on. When it's crashing it seems to take the shell down with it. And nothing is written to stdout nor stderr. So this might need more investigation.

**Comment by HertzDevil** [comment]
I tried the above on Windows 11 22H2 22621.1555 and it never crashed.

Can `Process.new` simply insert _another_ named pipe in-between if it finds that an `IO` is non-blocking?

**Comment by straight-shoota** [comment]
Hm, after more digging it seems that the compiler (or the wrapper script?) is actually breaking. Not the program itself (I was using `run` and `eval` so it wasn't clear what's happening). I don't know what's going on there but it seems unrelated. So we're probably good on this.

@kubo Do you have any comment? Maybe we need to consider something we have missed so far?

**Comment by kubo** [comment]
> @kubo Do you have any comment? Maybe we need to consider something we have missed so far?

I checked it using [this gist code](https://gist.github.com/kubo/83ab504fc3243cd2e6ecee113f8874dc).
It seems that I made a mistake. Perhaps it was fixed by MS or probably I wrote wrong code.

As far as I ran the gist code, asynchronous **pipe** handles can be passed to child processes as stdio handles. `fgets` and `fputs` work fine in the child processes. However asynchronous **file** handles cannot. `fgets` fails with errno 22 (Invalid argument). `fputs` doesn't return error but writes nothing.

File: spec/std/http/spec_helper.cr (hunk 2 above)
Comment (straight-shoota) [reply spec/std/http/spec_helper.cr:84]: Yeah, this should require some investigation. The close behaviour should not be platform-specific.

Comment (HertzDevil) [reply spec/std/http/spec_helper.cr:84]: It looks like the current fiber is unconditionally scheduled at a later time even if `WriteFile` finishes synchronously, despite `FILE_FLAG_OVERLAPPED` being used.

Comment (HertzDevil) [reply spec/std/http/spec_helper.cr]: Now I'm wondering: if socket operations like `WSASend` [complete immediately](https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsasend#overlapped-socket-i-o), we also might not need to unconditionally reschedule inside `#wsa_overlapped_operation` either

Comment (straight-shoota) [reply spec/std/http/spec_helper.cr]: Great find! That seems like the piece we've been missing 🚀

**Review by straight-shoota — APPROVED** [review]
LGTM

**Pull request merged**